numpy/core/src/multiarray/methods.c

```diff
@@ -2092,16 +2090,26 @@ array_setstate(PyArrayObject *self, PyObject *args)
                                &(fa->flags));
     }
 
+    if (nbytes == 0) {
+        /* Nothing more to do, it should be OK to use the array data as-is */
+        if (!PyDataType_FLAGCHK(typecode, NPY_LIST_PICKLE)) {
+            Py_DECREF(rawdata);
+        }
+        /* Update flags. TODO: not all (maybe none) need updating */
+        PyArray_UpdateFlags(self, NPY_ARRAY_UPDATE_ALL);
+        Py_RETURN_NONE;
+    }
+    else {
+        /* Otherwise the data is already cleared */
+        assert(!PyArray_CHKFLAGS(self, NPY_ARRAY_OWNDATA));
+    }
+
```

Comment:
The changes here are the most complicated maybe.  The old code never did much if `num == 0`, but there is a problem that the flags were cleared above.  This removed the `OWNDATA` flag which in turn meant that `memhandler` was never decref'd (the array didn't even own any data!).

Moving this around means that we just keep the original data in this case, whatever that may be (probably also not quite correct!), I am not yet sure if/what it will do to those "UHOH mismatched..." prints.

Reply:
Using `Py_CLEAR` below for the memhandler, because it seems better to clear it, if we have no intention to decref it (since the array does not own its data).